cfme/markers/env_markers/provider.py

```diff
@@ -107,63 +231,125 @@ def providers(metafunc, filters=None, selector=ALL):
         flags_filter = ProviderFilter(required_flags=test_flags)
         filters = filters + [flags_filter]
 
-    potential_providers = list_providers(filters)
+    # available_providers are the ones "available" from the yamls after all of the global and
+    # local filters have been applied. It will be a list of crud objects.
+    available_providers = list_providers(filters)
+
+    # supported_providers are the ones "supported" in the supportability.yaml file. It will
+    # be a list of DataProvider objects and will be filtered based upon what the test has asked for
+    supported_providers = all_required('5.9', filters)
+
+    def get_valid_provider(provider):
+        # We now search theough all the available providers looking for one that matches the
+        # criteria. If we don't find one, we return None
+        for a_prov in available_providers:
+            try:
+                assert a_prov.version
+                if (a_prov.version == provider.version and
+                        a_prov.type == provider.type_name and
+                        a_prov.category == provider.category):
+                    return a_prov
+            except (AssertionError, KeyError):
```

Comment:
**required** this catch looks problematic, we should never use ASsertion errors for flow control

**question** where is `KeyError` from

Reply:
agreed, replaced AssertError.
I suspect KeyError can come from supportability.yaml.